File: python/완전탐색/소수 찾기.py

""" Level 2. 소수 찾기 """

from itertools import permutations
import math

def solution(numbers):

    answer = set()

    for count in range(1, len(numbers)+1):
        for num in list(permutations(numbers, count)):
    
            num = int("".join(num))
            if num not in answer and is_prime(num):
                answer.add(num)

    return len(answer)

def is_prime(number):

    if number < 2:
        return False
    
    for check in range(2, int(math.sqrt(number))+1):
        if number % check == 0:
            return False
        
    return True

print(solution("17"))       # 3
print(solution("011"))      # 2

File: Java/src/완전탐색/소수찾기.java
package 완전탐색;

import java.util.*;

public class 소수찾기 {

    static int N;
    static Set<Integer> numberSet;
    static boolean[] selected;

    public static void main(String[] args) {

        String n1 = "17";
        System.out.println(solution(n1));

        String n2 = "011";
        System.out.println(solution(n2));
    }

    public static int solution(String numbers) {
        
        N = numbers.length();
        
        numberSet = new HashSet<>();
        selected = new boolean[N];
        
        dfs(numbers, 0, "");
        
        int answer = 0;
        for (Integer num : numberSet) {
            if (1 < num && isPrime(num)) {
                answer += 1;
            }
        }
        
        return answer;
    }
    
    public static void dfs(String numbers, int count, String word) {
        
        if (count == N) {
            return;
        }
        
        for (int i = 0; i < N; i++) {
            if (!selected[i]) {
                selected[i] = true;
                numberSet.add(Integer.parseInt(word + numbers.charAt(i)));
                dfs(numbers, count+1, word + numbers.charAt(i));
                selected[i] = false;
            }
        }
    }
    
    public static boolean isPrime(int number) {
        
        for (int i = 2; i <= (int) Math.sqrt(number); i++) {
            if (number % i == 0) {
                return false;
            }
        }
        
        return true;
    }
}